---
title: "Manhattan Plot"
subtitle: "SNPs"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    code_folding: show
    theme: sandstone
---




### Theme
```{r}
whitetheme <- function() {
  theme_bw()+theme(
    plot.margin = margin(20,10,10,10),
    text = element_text(family="Helvetica", colour="black", face="plain"),
    
    plot.title = element_text(size=13, face="bold"),
    plot.subtitle = element_text(size=11, colour="black"),
    plot.caption = element_text(size=9, colour="black"),
    
    # PANEL
    panel.border = element_rect(fill=NA, colour = "black", size=0.7),
    
    # LEGEND
    legend.text = element_text(size=11),
    legend.title = element_blank(),
    legend.position = "top",
    legend.background = element_blank(),
    
    #AXIS
    axis.text.x=element_text(size=11, colour="black", face="plain"),
    axis.text.y = element_text(size=11, colour="black", face="plain"),
    axis.title.y = element_text(vjust=2),
    axis.title.x = element_text(vjust=-0.5),
    panel.grid=element_blank(),
    axis.line = element_line(color="black", size=0.25)
  )
}


```


### Packages and data
```{r include=FALSE}
library(ggplot2)


gwas.raw <- read.csv("results.csv") # Replace with path to already filtered SNPs
```


### Prepare data
```{r}
# Creat dataframe for representation
gwas <- data.frame(snp=gwas.raw$SNP, chr=gwas.raw$Chromosome, pos=gwas.raw$Position, pval=gwas.raw$P.value)
# Order by chromosome number
gwas <- gwas[order(gwas$chr),]

nchr <- length(unique(gwas$chr))

# Calculate cumulative position in bp
gwas$posCum <- NA
s <- 0
nbp <- c()
chr.pos <- c()
for (i in unique(gwas$chr)){
  nbp[i] <- max(gwas[gwas$chr == i,]$pos)
  gwas[gwas$chr == i,"posCum"] <- gwas[gwas$chr == i,"pos"] + s
  s <- s + nbp[i]
  
  chr.pos[i] <- (min(gwas[gwas$chr == i,"posCum"]) + max(gwas[gwas$chr == i,"posCum"])) / 2
}

# Chromosome position dataframe
chr.pos <- data.frame(chr=unique(gwas$chr), pos=chr.pos)
# x-axis limits
x.limits <- c(min(gwas$posCum), max(gwas$posCum))
```

### Annotation
```{r}
require(ggrepel)
# Define SNPs to be annotated
snps <- unique(gwas$snp)
pval5 <- gwas[-log10(gwas$pval)>5.5,]
pval5 <- pval5[complete.cases(pval5$snp),]
snps <- pval5$snp[11:23]
snp.annotate <- gwas[gwas$snp %in% snps,]

##### REPLACE WITH PROPER ANNOTATION LABELS #####
snp.annotate$annotation <- snp.annotate$snp
```



### Plot
```{r}
# Black and white option
colors <- rep(c("#5a5a5a", "#b2b2b2"), ceiling(nchr / 2))
# Colorful option (comment out for black and white)
colors <- c("#AA4465", "#F7BFB4", "#A8A4E0", "#DAEDBD", "#7FB285")

# Horizontal line for p-value cutoff (-log10)
# Default 10^-4 >>> -log10(10^-4) = 4
hline = 4


plot <- ggplot(data=gwas, aes(x=posCum, y=-log10(pval), color=as.factor(chr)))+
          geom_point() +
          whitetheme()+
          scale_x_continuous(label=chr.pos$chr, breaks=chr.pos$pos, 
                             limits=x.limits, expand=c(0,0)) +
          scale_y_continuous(expand=c(0,0), limits=c(-.02,10.02), breaks=seq(0,10,1)) + 
          scale_color_manual(values=colors) +
          geom_hline(yintercept = hline, linetype = "dashed", lwd = 0.5)+
          labs(x="Chromosome",
               y=expression(paste("-log"["10"],"(",italic("p"), ")")))+
          theme(legend.position = "None")+ # Remove legend
          geom_label_repel(data=snp.annotate, aes(label=annotation), size=2)+
          geom_point(data=snp.annotate, color="black")
plot
```

### Save plot
```{r}
ggsave(filename = "manhattan.tiff",
       plot = plot,
       device = "tiff",
       width = width,
       height = height,
       units = "cm")
```